sync/backup.go

```diff
@@ -150,16 +153,132 @@ func (t *Task) restoreBackup(replicaInController *types.ControllerReplicaInfo, b
 		return err
 	}
+
 	return nil
 }
 
+func (*Task) IsRestorationComplete(rsList []*RestoreStatus) bool {
+	if rsList == nil {
+		return false
+	}
+	for _, r := range rsList {
+		if r.Progress != 100 {
+			return false
+		}
+	}
+	return true
+}
+
+func (t *Task) GetRestoreStatus() ([]*RestoreStatus, error) {
+	var restoreStatusList []*RestoreStatus
+	isIncrementalFullRestore := false
+	snapshotFile := ""
+	replicas, err := t.client.ReplicaList()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, r := range replicas {
+		repClient, err := replicaClient.NewReplicaClient(r.Address)
+		if err != nil {
+			logrus.Errorf("Cannot create a replica client for IP[%v]: %v", r.Address, err)
+			return nil, err
+		}
+
+		rsList, err := repClient.BackupRestoreStatus()
+		if err != nil {
+			return nil, err
+		}
+
+		if rsList.DestFileName == "" {
+			//No Active Backup Restore happening
+			continue
+		}
+
+		snapshotFile = rsList.DestFileName
+
+		restoreStatusList = append(restoreStatusList, &RestoreStatus{
+			Progress:     int(rsList.Progress),
+			RestoreError: rsList.Error,
+			SnapshotName: rsList.DestFileName,
+		})
+
+		//Check if this is incremental full restore
+		if rsList.Progress == 100 && strings.HasSuffix(snapshotFile, ".snap_tmp") {
+			isIncrementalFullRestore = true
+			tmpSnapshotDiskName := snapshotFile
+			snapshotDiskName, err := replica.GetSnapshotNameFromTempFileName(snapshotFile)
+			snapshotDiskMetaName := replica.GenerateSnapshotDiskMetaName(snapshotDiskName)
+			tmpSnapshotDiskMetaName := replica.GenerateSnapshotDiskMetaName(tmpSnapshotDiskName)
+
+			defer func() {
+				// try to cleanup tmp files
+				repClient.RemoveFile(tmpSnapshotDiskName)
+				repClient.RemoveFile(tmpSnapshotDiskMetaName)
+			}()
+
+			// replace old snapshot
+			if err = repClient.RenameFile(tmpSnapshotDiskName, snapshotDiskName); err != nil {
+				return nil, fmt.Errorf("failed to replace old snapshot %v with the fully restored file %v: %v", snapshotDiskName, tmpSnapshotDiskName, err)
+			}
+			if err = repClient.RenameFile(tmpSnapshotDiskMetaName, snapshotDiskMetaName); err != nil {
+				return nil, fmt.Errorf("failed to replace old snapshot meta %v with the fully restored meta file %v: %v", snapshotDiskMetaName, tmpSnapshotDiskMetaName, err)
+			}
+
+			// snapshot files get changed, need reload
+			_, err = repClient.ReloadReplica()
+			if err != nil {
+				logrus.Warnf("Failed to reload replica %v: %v", r.Address, err)
+			}
+		}
+	}
+
+	if len(restoreStatusList) == 0 {
+		//No Active Backup Restore happening
+		return nil, nil
+	}
+
+	// Not required to revert the snapshot for incremental full backup as it's renamed and replica is reloaded
+	if !isIncrementalFullRestore && t.IsRestorationComplete(restoreStatusList) {
+		snapshotID, err := replica.GetSnapshotNameFromDiskName(snapshotFile)
+		if err != nil {
+			return nil, err
+		}
+
+		// call to controller to revert to sfile
+		if err := t.client.VolumeRevert(snapshotID); err != nil {
```

Comment:
You cannot do functional change in a function suppose only to get the status. GetRestoreStatus is not required to be called at all. You need to have a thread (goroutine) waiting for the restoration completed then call the revert or rename. It needs to be run on the server side (since the client would be gone).

Reply:
Yes, Removing and Renaming activity can be done on the server side and I've making those changes and testing them. 

However, Volume Revert and Replica Reload definitions are implemented in Controller Server and Replica Server respectively. I don't know how I can call those functions from SyncAgentServer as we don't have controller and replica client. Even to create those clients, I'd need some information which requires to be sent to SyncAgentServer (which I don't think is the right thing to do)

So is it okay to keep it in sync backup?

Reply:
No, you cannot do this. This status check shouldn't have any side effect. Also, just imagine if the user never calls to check the status of restoration, then the restoration will never be completed.

But the point you raised is valid. There is a way out. During the restoration, if no frontend is enabled, the revert won't be needed in the controller since it's mainly used to refresh the cache of the block device (frontend). Also, the currently logic makes replicas automatically revert to the restored snapshot, which will cause problems anyway if the controller's frontend is still up.

So we can add a few constraints to make it work:
1. If the controller's frontend is up, no restoration is allowed.
2. If a restoration is in progress for any replica, `start-frontend` of the controller is not allowed. For this purpose, replicas need to reports if it's restoring or not.

Since replicas revert to the newly restored snapshot automatically, after the restoration is completed, the frontend can be started again and the volume should contain the restored volume content.